eth/rpc/events_parser.go

```diff
@@ -0,0 +1,169 @@
+			if err != nil {
+				return nil, err
+			}
+			ethTxIndexFromEvent, err := strconv.ParseUint(eventEthereumTx.Index, 10, 31)
+			if err != nil {
+				return nil, err
+			}
+			gasUsed, err := strconv.ParseUint(eventEthereumTx.GasUsed, 10, 64)
+			if err != nil {
+				return nil, err
+			}
+			committedTx := ParsedTx{
+				MsgIndex:   msgIndex,
+				EthTxIndex: int32(ethTxIndexFromEvent),
+				EthHash:    gethcommon.HexToHash(eventEthereumTx.EthHash),
+				GasUsed:    gasUsed,
+				Failed:     len(eventEthereumTx.EthTxFailed) > 0,
+			}
+			// replace pending tx with committed tx
+			if msgIndex >= 0 && len(parsedTxs.Txs) == msgIndex+1 {
+				parsedTxs.Txs[msgIndex] = committedTx
+			} else {
+				// EventEthereumTx without EventPendingEthereumTx
+				return nil, errors.New("EventEthereumTx without pending_ethereum_tx event")
+			}
+		}
+	}
+
+	// this could only happen if tx exceeds block gas limit
+	if result.Code != 0 && tx != nil {
+		for i := 0; i < len(parsedTxs.Txs); i++ {
+			parsedTxs.Txs[i].Failed = true
+
+			// replace gasUsed with gasLimit because that's what's actually deducted.
+			gasLimit := tx.GetMsgs()[i].(*evm.MsgEthereumTx).GetGas()
+			parsedTxs.Txs[i].GasUsed = gasLimit
+		}
+	}
+	return parsedTxs, nil
+}
+
+// ParseTxIndexerResult parse tm tx result to a format compatible with the custom tx indexer.
+func ParseTxIndexerResult(txResult *tmrpctypes.ResultTx, tx sdk.Tx, getter func(*ParsedTxs) *ParsedTx) (*eth.TxResult, error) {
+	txs, err := ParseTxResult(&txResult.TxResult, tx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse tx events: block %d, index %d, %v", txResult.Height, txResult.Index, err)
+	}
+
+	parsedTx := getter(txs)
+	if parsedTx == nil {
+		return nil, fmt.Errorf("ethereum tx not found in msgs: block %d, index %d", txResult.Height, txResult.Index)
+	}
+	index := uint32(parsedTx.MsgIndex) // #nosec G701
+	return &eth.TxResult{
+		Height:            txResult.Height,
+		TxIndex:           txResult.Index,
+		MsgIndex:          index,
+		EthTxIndex:        parsedTx.EthTxIndex,
+		Failed:            parsedTx.Failed,
+		GasUsed:           parsedTx.GasUsed,
+		CumulativeGasUsed: txs.AccumulativeGasUsed(parsedTx.MsgIndex),
+	}, nil
+}
+
+// GetTxByHash find ParsedTx by tx hash, returns nil if not exists.
+func (p *ParsedTxs) GetTxByHash(hash gethcommon.Hash) *ParsedTx {
+	if idx, ok := p.TxHashes[hash]; ok {
+		return &p.Txs[idx]
+	}
+	return nil
+}
+
+// GetTxByMsgIndex returns ParsedTx by msg index
+func (p *ParsedTxs) GetTxByMsgIndex(i int) *ParsedTx {
+	if i < 0 || i >= len(p.Txs) {
+		return nil
+	}
+	return &p.Txs[i]
+}
+
+// GetTxByTxIndex returns ParsedTx by tx index
+func (p *ParsedTxs) GetTxByTxIndex(txIndex int) *ParsedTx {
+	if len(p.Txs) == 0 {
+		return nil
+	}
+	// assuming the `EthTxIndex` increase continuously,
+	// convert TxIndex to MsgIndex by subtract the begin TxIndex.
+	msgIndex := txIndex - int(p.Txs[0].EthTxIndex)
```

Comment:
This claim is incorrect because the block 
```go
if len(p.Txs) == 0 {
		return nil
	}
```

is an empty check for `p.Txs` already. No issue here.